### PR TITLE
feat: add support PEP621 compatible pyproject.toml

### PR DESCRIPTION
**Pull request by jenshnielsen**
:heart: Thank you for your contribution!

### Description

This adds support for extracting requirements from pyproject.toml as defined by [pep621](https://peps.python.org/pep-0621/)
Setuptools 61 added support for this and setuptools 65 has stabilized this. Supporting this also means that requirements-builder will be useful for non setuptools projects that uses pep621

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [x] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [x] I've added relevant test cases. Tested to the same extend as setup.cfg
- [x] I've added relevant documentation. Neither setup.cfg nor pyproject.toml is explicitly documented not sure if we should change that.
- [x] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code). **NA**
- [x] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code). **NA**
- [x] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code). **NA**
- [X] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [ ] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions). I am the only contributer of the added code. Not sure if any changes to copyright headers are required. I do not need to be attributed.
- [ ] I've NOT included third-party code (copy/pasted source code or new dependencies). Added tomli dependency for parsing toml files on python <3.11

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

The architects team no longer exist (above link is a 404)

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

**Review by greut — APPROVED** [review]
It seems mostly okay to me; I'll work aside to add Poetry support.

File: tests/test_requirements_builder.py [hunk 1]
```diff
@@ -17,6 +18,11 @@
 SETUP = abspath(join(dirname(__file__), "./fixtures/setup.txt"))
 
 
+def _has_toml_lib():
+    # tomlib is part of the batteries included since 3.11
+    if sys.version_info >= (3, 11)
```
Comment (greut) [inline tests/test_requirements_builder.py:23]: In fact.

```suggestion
    return sys.version_info >= (3, 11)
```

**Comment by greut** [comment]
If you run _tox_ you'll get a coupe of errors, line length or stuff like that.

**Comment by jenshnielsen** [comment]
@greut  I cannot run your tox installation as it seems to pick a pytest / pytest-docstyle pytest-codestyle combination that is incompatible. 

```
Installed c:\users\jenielse\source\repos\requirements-builder\.eggs\pycodestyle-2.9.1-py3.8.egg
Traceback (most recent call last):
  File "setup.py", line 13, in <module>
    setup()
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\__init__.py", line 87, in setup
    return distutils.core.setup(**attrs)
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\_distutils\core.py", line 177, in setup
    return run_commands(dist)
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\_distutils\core.py", line 193, in run_commands
    dist.run_commands()
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\_distutils\dist.py", line 968, in run_commands
    self.run_command(cmd)
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\dist.py", line 1217, in run_command
    super().run_command(command)
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\_distutils\dist.py", line 987, in run_command
    cmd_obj.run()
  File "c:\users\jenielse\source\repos\requirements-builder\.eggs\pytest_runner-6.0.0-py3.8.egg\ptr\__init__.py", line 196, in run
    installed_dists = self.install_dists(dist)
  File "c:\users\jenielse\source\repos\requirements-builder\.eggs\pytest_runner-6.0.0-py3.8.egg\ptr\__init__.py", line 147, in install_dists
    orig.test.install_dists(dist), self.install_extra_dists(dist)
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\command\test.py", line 195, in install_dists
    tr_d = dist.fetch_build_eggs(dist.tests_require or [])
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\setuptools\dist.py", line 875, in fetch_build_eggs
    resolved_dists = pkg_resources.working_set.resolve(
  File "C:\Users\jenielse\source\repos\requirements-builder\.tox\py38\lib\site-packages\pkg_resources\__init__.py", line 800, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (pytest 6.2.5 (c:\users\jenielse\source\repos\requirements-builder\.eggs\pytest-6.2.5-py3.8.egg), Requirement.parse('pytest>=7.0'), {'pytest-pydocstyle', 'pytest-pycodestyle'})
```

However, I have run your tests manually and I think this should fix it

**Comment by jenshnielsen** [comment]
@greut This should pass the formatting tests. Would it be possible to merge?

**Review by jirikuncar — COMMENTED** [review]
We might need to support alternative build backends like Poetry too. It should be explicitly noted in release notes that this PR works only with `"setuptools.build_meta"` backend.

**Review by jirikuncar — APPROVED** [review]

**Comment by jenshnielsen** [comment]
@jirikuncar That is not the case. It works with every backend that supports [pep621](https://peps.python.org/pep-0621/) that should also include flit and hatch. But happy do document this if someone could point me to where that should go (support for setup.cfg seems almost undocumented)

**Comment by greut** [comment]
@jirikuncar there is the beginning of a poetry support around there, https://github.com/inveniosoftware/requirements-builder/tree/fix/poetry-support

**Comment by jirikuncar** [comment]
@jenshnielsen I have modified the PR title. We can extend https://github.com/inveniosoftware/requirements-builder/blob/master/requirements_builder/__init__.py#L10-L29 with updated `--help` output.

**Comment by greut** [comment]
As [poetry 1.2](https://python-poetry.org/blog/announcing-poetry-1.2.0/) is moving some lines, e.g. `dev-dependencies`. I think it's better to stick to PEP621 first.

**Pull request merged**